### PR TITLE
Fix for duplicatiing transactions (IS-869)

### PR DESCRIPTION
**Pull request by dimalit**
1. Skip invalid transactions from block, starting from "patch timestamp"
2. Ignore them in historic build, if they are present in block

Fixes https://github.com/skalenetwork/internal-support/issues/869

Tests:
1. test/unittests/libethereum/SkaleHost.cpp tests how various kinds of invalid transactions are skipped from block (both for enabled patch and for disabled patch)
2. test/unittests/libweb3jsonrpc/jsonrpc.cpp -> skip_invalid_transactions tests affected JSON-RPC calls on historic node

### CODE REVIEW

**Review by kladkogex — CHANGES_REQUESTED** [review]
Please specify in detail how the fix was tested.

Also please add comments to the code describing the fix

File: libskale/SkipInvalidTransactionsPatch.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+time_t SkipInvalidTransactionsPatch::activationTimestamp;
+time_t SkipInvalidTransactionsPatch::lastBlockTimestamp;
+
+bool SkipInvalidTransactionsPatch::activateTimestampPassed() {
+    if ( activationTimestamp == 0 )
+        return false;
+    return lastBlockTimestamp >= activationTimestamp;
+}
+
+bool SkipInvalidTransactionsPatch::needToKeepTransaction( const Transaction& tr, bool excepted ) {
```
Comment (kladkogex) [inline libskale/SkipInvalidTransactionsPatch.cpp]: Please remove tr and also use underscore for function variables

File: libethereum/Client.cpp [hunk 2]
```diff
@@ -654,7 +656,7 @@ size_t Client::syncTransactions(
     RevertableFSPatch::lastBlockTimestamp = blockChain().info().timestamp();
     StorageDestructionPatch::lastBlockTimestamp = blockChain().info().timestamp();
     POWCheckPatch::lastBlockTimestamp = blockChain().info().timestamp();
-
+    SkipInvalidTransactionsPatch::setLastBlock( blockChain().info() );
```
Comment (kladkogex) [inline libethereum/Client.cpp]: Please follow the logic of other patches  above - the code needs to be consistent

File: libskale/SkipInvalidTransactionsPatch.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+// 2) eth_getTransactionReceipt (affects "transactionIndex" field)
+// 3) eth_getBlockTransactionCountByHash/Number
+// 4) eth_getTransactionByHash (invalid transactions are treated as never present)
+// 5) eth_getTransactionByBlockHash/NumberAndIndex
+// Transactions are removed from Transaction Queue as usually.
+
+class SkipInvalidTransactionsPatch : public SchainPatch {
+public:
+    static void init( const dev::eth::ChainParams& cp ) {
+        activationTimestamp = cp.sChain.skipInvalidTransactionsPatchTimestamp;
```
Comment (kladkogex) [inline libskale/SkipInvalidTransactionsPatch.h]: Please add setter and print logs as other patch

**Review by kladkogex — CHANGES_REQUESTED** [review]
Please check the case if there is a single invalid transaction in the block.

**Review by kladkogex — CHANGES_REQUESTED** [review]
Please check the case if there is a single invalid transaction in the block.

File: libethereum/Client.cpp (hunk 2 above)
Comment (dimalit) [reply libethereum/Client.cpp]: It seems bad idea to stick just to timestamps - in general, patches can have different logic of activation. E.g. see AmsterdamFixPatch.

File: libskale/SkipInvalidTransactionsPatch.h (hunk 3 above)
Comment (dimalit) [reply libskale/SkipInvalidTransactionsPatch.h]: It seems bad idea to just stick to timestamp-only approach. Different patches can have different logic of activation. E.g. see AmsterdamFixPatch

**Comment by dimalit** [comment]
Case of single invalid transaction in block is checked in test/unittests/libethereum/SkaleHost.cpp

**Review by kladkogex — APPROVED** [review]

**Review by kladkogex — APPROVED** [review]

**Review by DmytroNazarenko — APPROVED** [review]

**Pull request merged**